### DISCUSSION TITLE
Topology optimization in MOOSE: Writing output of moose/modules/combined/examples/optimization/2d_mbb.i to a MSH or STL file?

### DISCUSSION THREAD

**Discussion by kurtenkera** [idaholab/moose#27378] [Q&A General]
I have run [moose/modules/combined/examples/optimization/2d_mbb.i](https://github.com/idaholab/moose/blob/next/modules/combined/examples/optimization/2d_mbb.i) and I'm curious to know if a general method exists to convert an outputted Exodus (.e) or VTK (.vtu) file of a topology-optimized 2D MBB component in MOOSE into a Gmsh (.msh) file or STL file? 

I have tried outputting the 2D MBB example first as a .vtk file in MOOSE, and then converting to a .msh file using [Meshio](https://github.com/nschloe/meshio) in Python. Below is the .vtk file of the optimized component in Paraview:

![image](https://github.com/idaholab/moose/assets/127472392/66cadd83-1a0e-4192-9b4b-313f5d970d77)

Below is the resulting .msh file in Gmsh (created by converting the .vtk file to a .msh file w/ Meshio). As you can see, the .msh file is just the original rectangular domain, whereas I need it to exclude all of the void regions? Would I need to somehow create code that converts all element density values to 0 or 1, and then exclude all nodes in the mesh that are associated with void regions in the rectangular domain, before converting to an STL or MSH file type? Thanks in advance

![image](https://github.com/idaholab/moose/assets/127472392/fea5c755-3e4c-45a3-9001-513a4764d3df)

**Comment by GiudGiud** [comment]
> Would I need to somehow create code that converts all element density values to 0 or 1, and then exclude all nodes in the mesh that are associated with void regions in the rectangular domain, before converting to an STL or MSH file type? Thanks in advance

this sounds like a good idea.
You can create a simulation that does just this:
- load the output of the previous one
- use a coupledVarThresholdSubdomainModifier to change the subdomain of the elements you want to get rid of

Save the result
- load the output of that simulation
- use the BlockDeletionGenerator to delete the elements in that new subdomain of the elements to get rid of

you can probably shorten this by executing `coupledVarThresholdSubdomainModifier` user object on FINAL at the end of your first simulation

**Comment by maxnezdyur** [comment]
There is a way to do this in Paraview. I don't remember the exact workflow but it's something like this. Use `iso volume` to threshold the density, then use the `extract surface,` then export as STL. I think the `extract surface` is needed but it's been a while.

**Reply by maxnezdyur** [reply]
Use `cell data to point data` to convert density to nodal values.
